src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/MemoryBuilder`1.cs

```diff
@@ -15,66 +17,114 @@ namespace Microsoft.AspNetCore.Razor;
+        => _memory[.._length];
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public void Append(T item)
     {
-        var index = _index;
-
+        var index = _length;
         var memory = _memory;
+
         if ((uint)index < (uint)memory.Length)
         {
             memory.Span[index] = item;
-            _index = index + 1;
+            _length = index + 1;
         }
         else
         {
             AppendWithResize(item);
         }
     }
 
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Append(ReadOnlySpan<T> source)
+    {
+        var index = _length;
+        var memory = _memory;
+
+        if (source.Length == 1 && (uint)index < (uint)memory.Length)
+        {
+            memory.Span[index] = source[0];
+            _length = index + 1;
+        }
+        else
+        {
+            AppendWithResize(source);
+        }
+    }
+
     [MethodImpl(MethodImplOptions.NoInlining)]
     private void AppendWithResize(T item)
     {
-        Debug.Assert(_index == _memory.Length);
-        var index = _index;
+        Debug.Assert(_length == _memory.Length);
+        var index = _length;
         Grow(1);
         _memory.Span[index] = item;
-        _index = index + 1;
+        _length = index + 1;
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private void AppendWithResize(ReadOnlySpan<T> source)
+    {
+        if ((uint)(_length + source.Length) > (uint)_memory.Length)
+        {
+            Grow(_memory.Length - _length + source.Length);
+        }
+
+        source.CopyTo(_memory.Span[_length..]);
+        _length += source.Length;
     }
 
     private void Grow(int additionalCapacityRequired = 1)
     {
+        Debug.Assert(additionalCapacityRequired > 0);
+
         const int ArrayMaxLength = 0x7FFFFFC7; // same as Array.MaxLength
 
-        var nextCapacity = Math.Max(_memory.Length != 0 ? _memory.Length * 2 : 4, _memory.Length + additionalCapacityRequired);
+        // Double the size of the array. If it's currently empty, default to size 4.
+        var nextCapacity = Math.Max(
+            _memory.Length != 0 ? _memory.Length * 2 : 4,
+            _memory.Length + additionalCapacityRequired);
 
+        // If nextCapacity exceeds the possible length of an array, then we want to downgrade to
+        // either ArrayMaxLength, if that's large enough to hold an additional item, or
+        // _memory.Length + 1, if that's larger than ArrayMaxLength. Essentially, we don't want
+        // to simply clamp to ArrayMaxLength if that isn't actually large enough. Instead, if
+        // we've grown too large, we want to OOM when Rent is called below.
         if ((uint)nextCapacity > ArrayMaxLength)
         {
-            nextCapacity = Math.Max(Math.Max(_memory.Length + 1, ArrayMaxLength), _memory.Length);
+            nextCapacity = Math.Max(
+                Math.Max(_memory.Length + 1, ArrayMaxLength),
+                _memory.Length);
```

Comment:
Ok but why are we also doing `Math.Max(..., _memory.Length)`? Is that because `_memory.Length + 1` might be smaller than `_memory.Length` in case of overflow? Maybe the comment should say that. Then I guess the assert below won't hold though.

In other words, it seems we are searching for `max { _memory.Length + 1, ArrayMaxLength, _memory.Length }` but that seems equivalent to `max { _memory.Length + 1, ArrayMaxLength }` unless we want to handle the overflow I guess.

Reply:
Yeah, I was thinking of the overflow. The situation is definitely not likely, but it's _possible_ that `_memory.Length == int.MaxValue`. I'll update the comment.

Reply:
On second thought, I don't think `_memory.Length == int.MaxValue` is actually possible. `Grow(...)` will never produce a `_memory` that large. And if anyone ever tried to create a `MemoryBuilder<T>` with `initialCapacity: int.MaxValue`, it would OOM because `0x7FFFFFFF` is larger than the maximum .NET array length of `0x7FFFFFC7`. So, I think it's safe to just take your suggestion.

Reply:
so in practice what happens when we try a code sample like this:

```cs
for (long i = 0; i < (long)int.MaxValue + 1; i++)
{
    builder.Append(0);
}
```

would it overflow the way you have it?